src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp

```diff
@@ -97,20 +97,25 @@ void MetadataStorageFromPlainObjectStorageCreateDirectoryOperation::undo(std::un
 {
     auto metadata_object_key = createMetadataObjectKey(object_key_prefix, metadata_key_prefix);
 
-    if (write_finalized)
+    if (write_finalized || write_created)
```

Comment:
From what I understand, we need to remove the `parent_path` from the `path_map` when `write_created` too.

Can we add a new method `removePath` in `InMemoryDirectoryPathMap` to remove the entry?

The following is my suggestion:

```C++
struct InMemoryDirectoryPathMap
{
    /**
     * @return the number of removed entries
     */
    size_t removePath(const std::filesystem::path & path)
    {
        std::lock_guard lock(mutex);
        return map.erase(path);
    }
}

void MetadataStorageFromPlainObjectStorageCreateDirectoryOperation::undo(std::unique_lock<SharedMutex> &)
{    
    if (!write_finalized && !write_created)
        return;

    size_t erase_count = path_map.removePath(path.parent_path());
    if (erase_count)
    {
       // Should we use `auto & metric` here?
        auto metric = object_storage->getMetadataStorageMetrics().directory_map_size;
        CurrentMetrics::sub(metric, erase_count);
    }

    StoredObject stored_object{metadata_object_key.serialize(), path / PREFIX_PATH_FILE_NAME};
    if (write_finalized)
        object_storage->removeObject(stored_object);
    else // write_created
        object_storage->removeObjectIfExists(stored_object);
}
```

Reply:
Thanks, I'll address it.